### ISSUE TITLE
Maybe a trick hack for the problem of "OSError: MoviePy error: failed to read the first frame of video file nCoronaVirus.mp4." exception.

### ISSUE THREAD

**Issue by cyberpsyche** [Zulko/moviepy#1078]
When I handle some video loop processing that moviepy will raise some kind of exception just like: 

> "OSError: MoviePy error: failed to read the first frame of video file nCoronaVirus.mp4. That might mean that the file is corrupted. That may also mean that you are using a deprecated version of FFMPEG. On Ubuntu/Debian for instance the version in the repos is deprecated. Please update to a recent version from the website." 

After I search on the internet I noticed that many people meet the same problem just like me. Even though that the exception message mentioned we should update deprecated version of FFMPEG, the installment of latest version of ffmpeg still cannot resolve the problem. 
 
So I looked into the source code of .../lib/python3.7/site-packages/moviepy/video/io/ffmpeg_reader.py and do a maybe-not-so-clean hack, but it does help a lot.

Here is the hack point, i.e. comment the last two line of the close() funciton:

```
def close(self):
        if self.proc:
            self.proc.terminate()
            self.proc.stdout.close()
            self.proc.stderr.close()
            self.proc.wait()
            self.proc = None
        #if hasattr(self, 'lastread'):
        #    del self.lastread 
```

I wonder if there is a better way to resolve this exception problem, I hope to be noticed if any buddy who could patch this code in a better way. And I also hope that the source maintainers could pay some attention to this issue, for it's maybe not really a problem related to ffmpeg but the inner logic of the ffmpeg_reader.py code. 

Looking forward to the update of moviepy source. Thanks the team. ;-)

**Comment by siddharthshah3030** [comment]
@cyberpsyche 
facing the same issue here.
it happens rarely, and I can't reproduce it.

can you share steps to reproduce this ?

**Comment by LauraLaureus** [comment]
I have the same error, as well as @siddharthshah3030 stated its a problem that is not happens always. I think there is something weird with subprocesses handling that causes this and the way to go on it's ignoring the problem as @cyberpsyche suggested. 

But I my case I would like to avoid "making a custom version of moviepy" since it's a dependecy of the project and I would like it to be easy-to-install. So I'll investigate on making overrides in my side to the problematic methods with some hope this problem disappears on version 2 of moviepy.

**Comment by mchulet** [comment]
Thanks for the workaround @cyberpsyche , I followed the steps and was able to overcome this error. Video generated properly after this fix, not sure if there are any side effects at the moment.

More info on what could be the cause would be of great help. 

In my case the error was coming for videos created using Vivo mobile phone, videos from other phones worked fine. Not sure whats missing from videos generated from Vivo phones.

**Comment by tburrows13** [comment]
Hey all, I've investigated a little, and I agree that there could be a bug in MoviePy's logic that causes the error to show when it is not supposed to. Note that if the file is genuinely corrupted, I would expect this error to show, however I think that it Is likely that ffmpeg would catch it first. However, the bugged logic only surfaces when the following warning is (legitimately, I think) raised:

```
"Warning: in file %s, " % (self.filename)
+ "%d bytes wanted but %d bytes read," % (nbytes, len(s))
+ "at frame %d/%d, at time %.02f/%.02f sec. "
% (self.pos, self.nframes, 1.0 * self.pos / self.fps, self.duration)
+ "Using the last valid frame instead."
```

So that is why you do not see it for every file, only for those files that are _slightly_ corrupted (but MoviePy can usually work around this by using the last valid frame as it says).

On to why `self.lastread` is not set:

It gets set in the `__init__` and the only place that it is deleted is in `.close()`. This is why removing the line that deletes it causes the error to go away. However, I'm reluctant to just do that and move on, because I want to know why the reader gets closed in the first place.
Somehow, the reader's `.close()` method is being called, which deletes both `reader.proc` and `reader.lastread`. Then when `reader.get_frame()` is next called, the proc is detected as not existing, so `reader.initialize()` and `reader.read_frame()` is called _without_ setting `reader.lastread`. Then when `read_frame()` can't find enough bytes in the output, the error is raised because `reader.lastread` does not exist.

So, why do we allow reinitialising of `proc`?https://github.com/Zulko/moviepy/blob/5c7a69ef72f7612766d8e1817d39b6abb29f2e08/moviepy/video/io/ffmpeg_reader.py#L193-L196
Why don't we instead raise an error there saying that you cannot read frames from a closed file.

The only reason that I can think of why the reader is being prematurely closed is that when a clip is copied (ie whenever it is transformed), both the copy and the original reference the same reader object. When the original is closed, this also calls `.close()` on the reader, which triggers what is described above. However, that would require the user to be calling `.close()` on the original clip. People who have experienced this problem, is this what you've been doing?

If this is the problem, then it probably requires of how MoviePy handles copying clips and their readers.

**Comment by tburrows13** [comment]
Update, I've found the source of the bug. It is not because of copys of the clip being closed as I suspected above (although that would also cause it), but it occurs when a frame that is out of bounds is requested _and_ when that frame is over 100 frames after the previously requested frame. I've included a fix in #1220 (which is mostly adding a `delete_lastread` parameter to `FFMPEG_VideoReader.close()` and setting that to False when reinitialising the clip). However, that PR also includes some improvements to make the logic more robust for the future.

This issue can be clearly seen in the test case that I have added `test_seeking_beyond_file_end`, in particular the second half of that test.
